File: internal/tool/tool_test.go
package tool

import (
	"fmt"
	"strings"
	"testing"

	"github.com/stretchr/testify/assert"
)

func TestTool_Execute(t *testing.T) {
	tests := []struct {
		name           string
		args           []string
		expectSuccess  bool
		expectOutput   string
		containsOutput string
	}{
		{
			name:          "executes simple command successfully",
			args:          []string{"echo", "hello"},
			expectSuccess: true,
			expectOutput:  "hello",
		},
		{
			name:          "executes command with multiple arguments",
			args:          []string{"echo", "hello", "world"},
			expectSuccess: true,
			expectOutput:  "hello world",
		},
		{
			name:           "captures stderr output",
			args:           []string{"sh", "-c", "echo 'error message' >&2"},
			expectSuccess:  true,
			containsOutput: "error message",
		},
		{
			name:          "handles command failure",
			args:          []string{"false"},
			expectSuccess: false,
			expectOutput:  "",
		},
		{
			name:           "handles non-existent command",
			args:           []string{"this-command-does-not-exist-12345"},
			expectSuccess:  false,
			containsOutput: "Studio error:",
		},
		{
			name:          "handles empty command",
			args:          []string{""},
			expectSuccess: false,
			expectOutput:  "Studio error: Empty command provided",
		},
		{
			name:          "handles command with spaces",
			args:          []string{"echo", "hello world with spaces"},
			expectSuccess: true,
			expectOutput:  "hello world with spaces",
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			result, err := Execute(tt.args[0], tt.args[1:]...)
			assert.NoError(t, err)
			assert.Equal(t, tt.expectSuccess, result.Success)

			if tt.expectOutput != "" {
				assert.Equal(t, tt.expectOutput, strings.TrimSpace(result.Output))
			}

			if tt.containsOutput != "" {
				assert.Contains(t, result.Output, tt.containsOutput)
			}
		})
	}
}

func TestTool_DebugMode(t *testing.T) {
	t.Run("debug mode is off by default", func(t *testing.T) {
		assert.False(t, IsDebugMode())
	})

	t.Run("can enable debug mode", func(t *testing.T) {
		SetDebugMode(true)
		assert.True(t, IsDebugMode())

		// Reset for other tests
		SetDebugMode(false)
		assert.False(t, IsDebugMode())
	})
}

func TestTool_CreateToolFunction(t *testing.T) {
	tests := []struct {
		name           string
		blueprint      Blueprint
		args           map[string]interface{}
		expectText     string
		expectIsError  bool
		expectContains string
	}{
		{
			name:       "creates function for simple command",
			blueprint:  &MockBlueprint{commandArgs: []string{"echo", "hello"}},
			args:       map[string]interface{}{},
			expectText: "hello",
		},
		{
			name:       "creates function with template arguments",
			blueprint:  &MockBlueprint{commandArgs: []string{"echo", "Hello World"}},
			args:       map[string]interface{}{"message": "Hello World"},
			expectText: "Hello World",
		},
		{
			name:          "handles command errors",
			blueprint:     &MockBlueprint{commandArgs: []string{"false"}},
			args:          map[string]interface{}{},
			expectText:    "",
			expectIsError: true,
		},
		{
			name:           "handles blueprint validation errors",
			blueprint:      &MockBlueprintWithError{err: fmt.Errorf("missing required parameter: name")},
			args:           map[string]interface{}{},
			expectIsError:  true,
			expectContains: "Validation error: missing required parameter: name",
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			fn := CreateToolFunction(tt.blueprint)
			result := fn(tt.args)

			assert.Len(t, result.Content, 1)
			assert.Equal(t, "text", result.Content[0]["type"])

			text := result.Content[0]["text"].(string)

			if tt.expectContains != "" {
				assert.Contains(t, text, tt.expectContains)
			} else {
				assert.Equal(t, tt.expectText, text)
			}

			assert.Equal(t, tt.expectIsError, result.IsError)
		})
	}
}

// MockBlueprint is a test helper that implements the Blueprint interface
type MockBlueprint struct {
	commandArgs []string
}

func (m *MockBlueprint) BuildCommandArgs(args map[string]interface{}) ([]string, error) {
	return m.commandArgs, nil
}

// MockBlueprintWithError is a test helper that returns an error
type MockBlueprintWithError struct {
	err error
}

func (m *MockBlueprintWithError) BuildCommandArgs(args map[string]interface{}) ([]string, error) {
	return nil, m.err
}

File: test/tool.test.ts
import { Tool } from '../src/tool';
import { Blueprint } from '../src/blueprint';

describe('Tool', () => {
  describe('execute', () => {
    it('executes simple command successfully', async () => {
      const { output, success } = await Tool.execute('echo', 'hello');
      expect(success).toBe(true);
      expect(output).toBe('hello');
    });

    it('handles command failure', async () => {
      const { output, success } = await Tool.execute('false');
      expect(success).toBe(false);
      expect(output).toBe('');
    });

    it('handles nonexistent command', async () => {
      const { output, success } = await Tool.execute('nonexistent-command-12345');
      expect(success).toBe(false);
      expect(output).toContain('Studio error:');
    });

    it('handles command with arguments', async () => {
      const { output, success } = await Tool.execute('echo', 'hello', 'world');
      expect(success).toBe(true);
      expect(output).toBe('hello world');
    });

    it('captures stderr output', async () => {
      const { output, success } = await Tool.execute('sh', '-c', 'echo "error" >&2');
      expect(success).toBe(true);
      expect(output).toContain('error');
    });

    it('handles empty command', async () => {
      const { output, success } = await Tool.execute('');
      expect(success).toBe(false);
      expect(output).toContain('Studio error:');
    });
  });

  describe('createToolFunction', () => {
    it('creates a tool function that executes commands', async () => {
      const blueprint = Blueprint.argv(['echo', '{{text}}']);
      const toolFunction = Tool.createToolFunction(blueprint);

      const result = await toolFunction({ text: 'hello world' });

      expect(result.content).toHaveLength(1);
      expect(result.content[0].type).toBe('text');
      expect(result.content[0].text).toBe('hello world');
      expect(result.isError).toBe(false);
    });

    it('handles tool function errors', async () => {
      const blueprint = Blueprint.argv(['false']);
      const toolFunction = Tool.createToolFunction(blueprint);

      const result = await toolFunction({});

      expect(result.content).toHaveLength(1);
      expect(result.content[0].type).toBe('text');
      expect(result.isError).toBe(true);
    });

    it('handles array arguments in tool function', async () => {
      const blueprint = Blueprint.argv(['echo', '[args...]']);
      const toolFunction = Tool.createToolFunction(blueprint);

      const result = await toolFunction({ args: ['hello', 'world', 'test'] });

      expect(result.content).toHaveLength(1);
      expect(result.content[0].type).toBe('text');
      expect(result.content[0].text).toBe('hello world test');
      expect(result.isError).toBe(false);
    });

    it('handles empty arguments in tool function', async () => {
      const blueprint = Blueprint.argv(['echo', '{{text}}']);
      const toolFunction = Tool.createToolFunction(blueprint);

      const result = await toolFunction({});

      expect(result.content).toHaveLength(1);
      expect(result.content[0].type).toBe('text');
      expect(result.content[0].text).toBe('');
      expect(result.isError).toBe(false);
    });
  });
});
